hands.go

```diff
@@ -0,0 +1,206 @@
+package goker
+
+import (
+	"reflect"
+	"sort"
+)
+
+// Hand - Represents a 5 card poker hand
+type Hand [5]Card
+
+// NewHand - Returns a pointer to a new hand consisting of
+// the required cards
+func NewHand(card1, card2, card3, card4, card5 Card) *Hand {
+	h := Hand([5]Card{card1, card2, card3, card4, card5})
+	sort.Sort(&h)
+	return &h
+}
+
+// NewHandFromSet - Returns a pointer to a new hand consisting of
+// the first five cards in the card set provided
+func NewHandFromSet(cards CardSet) *Hand {
+	return NewHand(cards[0], cards[1], cards[2], cards[3], cards[4])
+}
+
+// Rank - The relative value of the hand according to the rules
+// of poker
+func (h *Hand) Rank() HandRank {
+	if h.isFlush() && h.isStraight() {
+		if h.highCard().Rank() == Ace {
+			return newRoyalStraightFlush()
+		}
+		return newStraightFlush(h.highCard().Rank())
+	}
+
+	quads := h.groupsOf(4)
+	if len(quads) != 0 {
+		kicker := h.removeRanks(quads[0])[0]
+		return newFourOfAKind(quads[0], kicker.Rank())
+	}
+
+	trips := h.groupsOf(3)
+	pairs := h.groupsOf(2)
+
+	if len(trips) == 1 && len(pairs) == 1 {
+		return newFullHouse(trips[0], pairs[0])
+	}
+
+	if h.isFlush() {
+		return newFlush(h.ranks())
+	}
+
+	if h.isStraight() {
+		return newStraight(h.highCard().Rank())
+	}
+
+	if len(trips) == 1 {
+		otherCardSet := h.removeRanks(trips[0])
+		return newThreeOfAKind(trips[0], ranks(otherCardSet))
+	}
+
+	if len(pairs) == 2 {
+		kicker := h.removeRanks(pairs...)[0]
+		return newTwoPair(pairs[0], pairs[1], kicker.Rank())
+	}
+
+	if len(pairs) == 1 {
+		kickers := h.removeRanks(pairs[0])
+		return newPair(pairs[0], ranks(kickers))
+	}
+
+	return newHighCard(h.ranks())
+}
+
+func (h *Hand) isFlush() bool {
+	suit := h[0].Suit()
+	for _, c := range h {
+		if c.Suit() != suit {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (h *Hand) isStraight() bool {
+	if h.isAceLowStraight() {
+		return true
+	}
+
+	lastRank := h[0].Rank()
+	for i, card := range h {
+		if i == 0 {
+			continue
+		}
+
+		if card.Rank() != lastRank+1 {
+			return false
+		}
+		lastRank = card.Rank()
+	}
+
+	return true
+}
+
+// Returns all ranks such that the hand has n cards of that rank
+// Examples: KKQQQ.groupsOf(2) -> [K], KKQQQ.groupsOf(3) -> [Q]
+//           TTAA7.groupsOf(2) -> [T, A]
```

Comment:
TTAA7 is not a valid hand, right?  Wouldn't it be 7TTAA, since hands are in sorted order?  Later, I think, you do some extra work to sort things.  Maybe it would be better if this just returned groups in descending rank order?  So, [A, T]?